src/wp-includes/option.php

```diff
@@ -364,6 +364,157 @@ function get_options( $options ) {
 	return $result;
 }
 
+/**
+ * Sets the autoload value for an option in the database.
+ *
+ * Autoloading too many options can lead to performance problems, especially if the options are not frequently used.
+ * This function allows modifying the autoload value for an option without changing the actual option value. This is
+ * for example recommended for plugin activation and deactivation hooks, to ensure any options exclusively used by the
+ * plugin which are generally autoloaded can be set to not autoload when the plugin is inactive.
+ *
+ * Alternatively, {@see wp_set_options_autoload()} can be used to set the autoload value for multiple options at once.
+ *
+ * @since 6.4.0
+ *
+ * @global wpdb $wpdb WordPress database abstraction object.
+ *
+ * @param string      $option   Name of the option. Expected to not be SQL-escaped.
+ * @param string|bool $autoload Autoload value to control whether to load the option when WordPress starts up.
+ *                              Accepts 'yes'|true to enable or 'no'|false to disable.
+ * @return bool True if the autoload value was modified, false otherwise.
+ */
+function wp_set_option_autoload( $option, $autoload ) {
+	global $wpdb;
+
+	$autoload = ( 'no' === $autoload || false === $autoload ) ? 'no' : 'yes';
+
+	/*
+	 * Determine the current autoload value for the option.
+	 * If the value is empty, it means the option does not exist in the database.
+	 * If the old and new value are the same, no need to update.
+	 */
+	$old_autoload = $wpdb->get_var( $wpdb->prepare( "SELECT autoload FROM $wpdb->options WHERE option_name = %s", $option ) );
+	if ( ! $old_autoload || $old_autoload === $autoload ) {
+		return false;
+	}
+
+	// Update the autoload value.
+	$result = $wpdb->update(
+		$wpdb->options,
+		array( 'autoload' => $autoload ),
+		array( 'option_name' => $option )
+	);
+	if ( ! $result ) {
+		return false;
+	}
+
+	// Update caches as necessary based on the autoload change.
```

Comment:
Why not just delete the cache and let it be regenerated the next time `wp_load_alloptions()` is called? The case accounted for seems to be direct calls to `wp_cache_get( 'alloptions' );`?

Reply:
@boonebgorges For `alloptions` this would probably work, however omitting this logic block would lead to problems with individual caches potentially becoming stale. For example, consider the following scenario:
* Option is stored in DB but not autoloaded.
* `wp_set_option_autoload()` is called, changing the autoload value to `yes`.
* A bit later, the option is deleted from the database.
* Without the logic in question, when calling `get_option()` for the option, it may return the stale value of the option from before its autoload value was changed. That is because the option would not have been deleted from its individual cache in `wp_set_option_autoload()`, and then `delete_option()` would not have deleted it either since it would have only deleted it from `alloptions`, per the autoload value being `yes`.

The above is the main reason this code is present. Additionally, by "migrating" the value between caches, it avoids unnecessary database lookups that could occur after an autoload value change.

Reply:
As a general principle, I think it's a good idea to minimize the number of points in the codebase where the cache is populated. And in particular, it feels strange to me that we would perform "manual" changes on the contents of the cache outside of the primary place where the cache is generally populated. So I guess what I'm suggesting is that we only delete here - either the `alloptions` cache or the individual option cache - and then allow the cache to be repopulated the next time the function is called. This suggestion negates your "unnecessary database lookups", but IMO this is a feature and not a bug, since a database lookup guarantees cache integrity.

Reply:
@boonebgorges Updated in https://github.com/WordPress/wordpress-develop/pull/5069/commits/e704bc8dbe33befb4de347a5824b2a0298d601cc.

For context, in the case where only values are set to `no`, I decided to still delete them individually from `alloptions` rather than deleting the entire `alloptions` cache, since in this case we're not setting actual values, and this is similarly done in existing code when calling `delete_option()`, so I think it makes sense to follow that pattern.